### PR TITLE
Use a handle-based implementation

### PR DESCRIPTION
**Pull request by asayers**

### CODE REVIEW

**Comment by mpickering** [comment]
Forgive me if this has already been answered. 

Why is the SlackSession needed in the `SlackHandle`?

I also can't get the `sendRichMessage` function to work. See the example I added to the branch. When running this bot, the first message is echoed and repeated forever but without any additional formatting. Could you please have a look at this?

**Comment by asayers** [comment]
> Why is the SlackSession needed in the SlackHandle?

It doesn't need to be in `SlackHandle`. Instead, we could have

```haskell
withSlackHandle :: SlackConfig -> ((SlackHandle, SlackSession) -> IO a) -> IO a
-- instead of   :: SlackConfig -> ( SlackHandle                -> IO a) -> IO a
```

But I suspect users will often end up passing it around everywhere with the `SlackHandle`, since they're useful together.

The same thing applies to the `SlackConfig` - it's not necessary to have it in `SlackHandle`, but it's (*possibly*) convenient to have it. (The config is probably less useful than the session, however.)

**Comment by mpickering** [comment]
Sorry that I wasn't clear. My point is that if the `SlackSession` is not in the `SlackHandle` then it can be put in the state part of `RWST` (for example) and the API can be implemented as before.

I am really not a fan of the `getNextEvent` API and I don't understand why it is necessary.

**Comment by asayers** [comment]
I don't mind whether `SlackSession` is in the handle, or whether we leave it to the user to put it in `RWST` or something - whichever you prefer.

As for `getNextEvent`, it's an important part of the change I'm proposing; I suppose I didn't give my motivations, so I'll try to do that now. Essentially, I think the current API is quite opinionated about how users should structure their programs, whereas the `SlackHandle` API is a bit more flexible. Here's an (only slightly contrived) example to illustrate this point:

Suppose you want to consume your slack events using the machinery from `pipes`. You have a `Consumer Event m ()` and you want to create a `Producer Event m ()` to drive it with.

With `getNextEvent` you can write:

```haskell
slackProducer :: MonadSlack m => Producer Event m ()
slackProducer = forever $ lift getNextEvent >>= yield

main :: IO ()
main = runSlack conf $ runEffect $ slackProducer >-> slackConsumer
```

With the current API, you can do it, but you have to sort of smuggle the events out of the run-loop:

```haskell
slackProducer :: MonadIO m => MVar Event -> Producer Event m ()
slackProducer smuggler = forever $ liftIO (readMVar smugger) >>= yield

main :: IO ()
main = do
    smuggler <- newEmptyMVar
    let bot ev = liftIO (putMVar smuggler ev)
    void $ forkIO $ runBot conf bot ()
    runEffect $ slackProducer smuggler >-> slackConsumer
```

This feels like a work-around to me, and I can't think of a better way to do it.

(My use-case is actually different to this - I need to have multiple threads using the same `SlackHandle` - but I thought this would make a more succinct example.

**Comment by mpickering** [comment]
OK - that makes sense. I agree it is a good change to allow such flexibility. I would be happy if you pulled out the SlackSession from the SlackHandle.

Did you investigate the problem with `sendRichMessage`? That is currently the thing blocking this.

**Comment by asayers** [comment]
Yes, I'm looking at it now.

**Comment by asayers** [comment]
The reason it was repeating the message is because, unlike `sendMessage`, `sendRichMessage` triggers a `Message` event which you need to filter out by looking for your own `UserId`. I think this is just a quirk of the slack web API.

**Comment by asayers** [comment]
BTW, how do you feel about force-pushing to `master`? If you're ok with it, would you mind if I squashed ae3948e?

**Comment by mpickering** [comment]
Please add a comment to `sendRichMessage` about this. I think it's unlikely to trip people up too often as usually you filter the message input looking for specific triggers but it certainly is pitfall.

**Comment by mpickering** [comment]
I don't really mind about the commit history being very clean so I think we should leave it.

**Comment by wamaral** [comment]
Is this being considered yet? I forked this lib based on this branch, wondering if I could submit enhancements on top of it

**Comment by mpickering** [comment]
It is being considered. I am not convinced that the interface that is provided it very good but if people tell me that it is the interface that they want then I am happy to adopt it. 

If you read this comment thread and summarise whether you think all concerns have been addressed then I can merge these patches. 

I don't use this library at all so my maintenance is really just light touches to keep things building.

**Comment by asayers** [comment]
Sorry, I forgot to follow up on this. If I recall correctly, the weird thing about this branch is that it provides two interfaces. They're equivalently expressive, and it's up to the user which one to use. This could be quite confusing for people reading the docs for the first time. The other issue is that it's a breaking change.

But I *do* this we should get this merged in some form, since it's strictly more flexible than the existing API (try implementing the [TimerBot example](https://github.com/mpickering/slack-api/blob/monad-slack/example/TimerBot.hs), for instance).

**Comment by wamaral** [comment]
The project is at 0.x so breaking changes are to be expected. However that's still potentially frustrating for current customers.

Regarding the interface, I couldn't say it's the best choice, but for my use case (piping `Event`s through `STM Chan`s) it definitely felt more accessible.

About `sendRichMessage`, first of all it's good to have a proper function to call. The "repeating event" is actually a quite common issue with bots, you should always be mindful about filtering the bot's own messages. I don't think it should be the function's concern (it might even be desirable to react to self in some circumstances)

**Comment by mpickering** [comment]
@wamaral I think the concern is that there is another function `sendMessage` which is a RTM API function which doesn't have this problem. I don't think it is desirable for `sendRichMessage` to behave very differently to `sendMessage`?

File: README.md [hunk 1]
```diff
@@ -1,35 +1,29 @@
-![Travis Build Status](https://travis-ci.org/mpickering/slack-api.svg?branch=master)
+Bindings to the Slack RTM API. More information can be found [here](https://api.slack.com/rtm).
```
Comment (asayers) [inline README.md]: These README changes seem superfluous.

**Comment by wamaral** [comment]
@mpickering I see, that makes sense. Yeah, both functions should optimally behave the same.

In that case, what if we implement `sendMessage` in terms of `sendRichMessage`? Here's how I use in my code, I ignore `sendMessage` and use `sendRichMessage` for regular messages:

```haskell
publishMessage :: OutputResponse -> Slack ()
publishMessage resp = do
  let cid = channel $ event resp
  case (message resp) of
    SimpleMessage txt -> slackSendMsg cid txt []
    RichMessage rmsg  -> slackSendMsg cid "" [rmsg]
    NoMessage         -> pure ()

slackSendMsg :: ChannelId -> T.Text -> [Attachment] -> Slack ()
slackSendMsg cid msg att = do
  send <- sendRichMessage cid msg att
  either (liftIO . putStrLn . T.unpack) pure send
```

So `sendRichMessage` with an empty `[Attachment]` generates the same output as `sendMessage`, while still triggering a new `Event`

**Comment by asayers** [comment]
I rebased on v0.12 and added a note to the `sendRichMessage` docs about the different behaviour.

**Comment by asayers** [comment]
The addition of `sendRichMessage` is really independent from the API rework, so if this issue is blocking then we could split the PR in two.

**Comment by mitchellwrosen** [comment]
@asayers Hello! I'd like to suggest `pipes-concurrency` as a possible solution to your problem. The trick is to express the entire `runBot` invocation as your `Producer Event`. Here's an example:

```haskell
import Control.Concurrent.Async
import Control.Monad.Managed
import Pipes
import Pipes.Concurrent
import Web.Slack

main :: IO ()
main = runManaged (runEffect (slackEvents >-> echoConsumer))

-- The producer - all Slack events
slackEvents :: MonadManaged m => SlackConfig -> Producer Event m ()
slackEvents config = do
  -- Spawn an unbounded pipe
  (out, inp) <- liftIO (spawn unbounded)

  -- For each event, write it to the pipe
  let callback :: Event -> IO ()
      callback event = do
        _ <- atomically (send out event)
        pure ()
  
  -- Run 'runBot' in a background thread
  (lift . using . managed . withAsync)
    (runBot config (liftIO . callback) ())

  -- Yield each event written to the pipe by 'callback'
  fromInput inp

-- The consumer - just echo
echoConsumer :: MonadIO m => Consumer Event m a
echoConsumer = for cat (liftIO . print)
```

Note that you have to throw away the `SlackState` using this approach (which is required anyways, when moving to a multi-threaded model). You can recover it if you'd like using whatever IORef/MVar/TVar makes sense.

I admit I agree with @mpickering, I'm not a huge fan of the new `getNextEvent` API. (However, I'm also not a huge fan of the existing `StateT` + callback API, either :))

**Comment by asayers** [comment]
Hi! You've identified the fact that the current API forces you to use multiple threads to implement something like `slackEvents` - removing this limitation is precisely the motivation for this PR! (See https://github.com/mpickering/slack-api/pull/68#issuecomment-243090375)

It may not be obvious at first glance, but the API provided by `Web.Slack.Handle` is in fact a pipes/managed sort of API. (`getNextEvent` is `await` by another name, and `withSlackHandle` is the kind of pattern `Managed` wraps.) To illustrate this, let's do the conversion and implement your `slackEvents` function:

```haskell
slackEvents :: MonadManaged m => SlackConfig -> Producer Event m a
slackEvents cfg = do
    h <- lift $ using $ managed $ withSlackHandle cfg
    forever $ liftIO (getNextEvent h) >>= yield
```

I'll admit, `slackEvents` is a very tidy abstraction - so long as the only thing you do with your slack session is receive events. If you also want to send messages with the same session, I think you'll end up exposing `SlackHandle` to the user with something like

```haskell
slackEvents :: MonadIO m => SlackHandle -> Producer Event m a
slackEvents h = forever $ liftIO (getNextEvent h) >>= yield

msgConsumer :: MonadIO m => SlackHandle -> Consumer (ChannelId, Text) m a
msgConsumer h = forever $ await >>= liftIO . uncurry (sendMessage h)
```

**Comment by mitchellwrosen** [comment]
Oh yeah! I totally neglected to consider sending events to Slack from a bot. That seems unreasonably difficult to express with the current API (I assume you would smuggle the `Connection` out of the callback to the `Hello` event).

**Comment by asayers** [comment]
The latest version of this branch is simplified, exposing only the `SlackHandle`-based API. I [uploaded the docs](http://www.asayers.com/slack-api/Web-Slack.html) to make it easier to review.

**Comment by mpickering** [comment]
I am really deferring to the likes of @mitchellwrosen who seems to actually use the library.

**Comment by mitchellwrosen** [comment]
Looks pretty good to me. Is the `Web.Slack.Compat` module necessary?

**Comment by mitchellwrosen** [comment]
Also: can `sendMessage` send to channels, groups, and individual users with that type? Or does the ID need to be something more general?

**Comment by asayers** [comment]
> can sendMessage send to channels, groups, and individual users with that type?

The slack API supports sending messages to regular/group channels (which we represent as `ChannelId`) or DM channels (which we represent as `IMId`).; currently, Web.Slack exports these:

```
imToChannel :: IMId -> ChannelId
channelToIM :: ChannelId -> IMId
```

I agree, this is sub-optimal - ideally `send{,Rich}Message` would accept either of these types.

**Comment by asayers** [comment]
> Is the Web.Slack.Compat module necessary?

No, not really. Do you think the library would be cleaner without it? I guess so...

**Comment by mitchellwrosen** [comment]
Yeah, I'd prefer to just omit the module. A major version bump is sufficient for breaking-changes, no need to keep around the old API forever :)

**Comment by asayers** [comment]
I dropped the compatibility module and rebased on master.

**Comment by mitchellwrosen** [comment]
I rebased again onto master and tucked the executables under a `build-examples` flag. Could you please update then CHANGELOG for me @asayers? You know better that I what went into all this :)

Sorry it took so long to merge, I didn't really realize I was in charge of shepherding it thru :D

**Comment by asayers** [comment]
@mitchellwrosen I added a line to the changelog. It certainly wasn't on you to get this merged - it's my fault it's taken so long. I just threw this over the wall and didn't push it out of pure laziness.

**Pull request merged**